internal/refreshtoken/refresh_token.go

```diff
@@ -0,0 +1,156 @@
+// list endpoint refresh tokens. Refresh tokens are used when users
+// paginate through results in our list endpoints, and also to
+// allow users to request new, updated and deleted resources.
+package refreshtoken
+
+import (
+	"bytes"
+	"context"
+	"time"
+
+	"github.com/hashicorp/boundary/internal/boundary"
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/boundary/internal/types/resource"
+)
+
+// UpdatedTimeBuffer is used to automatically adjust the updated
+// time of a refresh token to account for delays between overlapping
+// database transactions.
+const UpdatedTimeBuffer = 30 * time.Second
+
+// A Token is returned in list endpoints for the purposes of pagination
+type Token struct {
+	CreatedTime         time.Time
+	UpdatedTime         time.Time
+	ResourceType        resource.Type
+	GrantsHash          []byte
+	LastItemId          string
+	LastItemUpdatedTime time.Time
+}
+
+// New creates a new refresh token from a createdTime, resource type, grants hash, and last item information
+func New(ctx context.Context, createdTime time.Time, updatedTime time.Time, typ resource.Type, grantsHash []byte, lastItemId string, lastItemUpdatedTime time.Time) (*Token, error) {
+	const op = "refreshtoken.New"
+
+	if len(grantsHash) == 0 {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "missing grants hash")
+	}
+	if createdTime.After(time.Now()) {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "created time is in the future")
+	}
+	if createdTime.Before(time.Now().AddDate(0, 0, -30)) {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "created time is too old")
+	}
+	if updatedTime.Before(createdTime) {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "updated time is older than created time")
+	}
+	if updatedTime.After(time.Now()) {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "updated time is in the future")
+	}
+	if lastItemId == "" {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "missing last item ID")
+	}
+	if lastItemUpdatedTime.After(time.Now()) {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "last item updated time is in the future")
+	}
+
+	return &Token{
+		CreatedTime:         createdTime,
+		UpdatedTime:         updatedTime,
+		ResourceType:        typ,
+		GrantsHash:          grantsHash,
+		LastItemId:          lastItemId,
+		LastItemUpdatedTime: lastItemUpdatedTime,
+	}, nil
+}
+
+// FromResource creates a new refresh token from a resource and grants hash.
+func FromResource(res boundary.Resource, grantsHash []byte) *Token {
+	t := time.Now()
+	return &Token{
+		CreatedTime:         t,
+		UpdatedTime:         t,
+		ResourceType:        res.GetResourceType(),
+		GrantsHash:          grantsHash,
+		LastItemId:          res.GetPublicId(),
+		LastItemUpdatedTime: res.GetUpdateTime().AsTime(),
+	}
+}
+
+// LastItem returns the last item stored in the token.
+func (rt *Token) LastItem() *item {
```

Comment:
Why is an exported method returning an unexported type?

Reply:
It prevents the type from being constructed other than via this function, but I'll export it since this was surprising.